locator/tablets.cc

```diff
@@ -616,9 +616,19 @@ future<bool> check_tablet_replica_shards(const tablet_metadata& tm, host_id this
 }
 
 class tablet_effective_replication_map : public effective_replication_map {
+    using replication_factor_t = uint16_t;
+    using replication_factor_list = utils::small_vector<replication_factor_t, 8>;
+
     table_id _table;
     tablet_sharder _sharder;
     mutable const tablet_map* _tmap = nullptr;
+    // The `_datacenter_map` stores the association between each data center
+    // and its index in the `replication_factor_list` in `_tablet_to_dc_replication_factor_map`.
+    absl::flat_hash_map<sstring, replication_factor_t> _datacenter_map;
+    // The `_tablet_to_dc_replication_factor_map` stores replication factors
+    // per data center for each Tablet.
+    std::vector<replication_factor_list> _tablet_to_dc_replication_factor_map;
```

Comment:
Should be chunked_vector to avoid large contiguous allocations.

Reply:
Can you please elaborate why is it bad to have a large allocation? Do you mean our allocator's storage might be already so fragmented that there would not be such a big chunk of memory (~5 Mb for 100k tablets ) ?

Reply:
Yes, the larger the allocation to more likely it is to cause premature OOM. In practice, we will try to reclaim contiguous space back from the LSA region (memtables/cache), which occupies higher addresses, but that may still require a lot of work to evict data and move segments around, and may cause reactor stalls. We maintain a reserve of free standard allocator space which is supposed to amortizes this cost, but if the request cannot be met due to fragmentation, it cannot make use of that reserve. 

As a general rule, we avoid contiguous allocations larger than 128K (LSA segment size). 128K is a unit of memory reclamation from LSA to the standard allocator's pool.

Reply:
In the benchmark I see that the number of tablets per table is 2048. Is that a typical number for production or it's just a testing number? Also when I set it to 65k I get an error "Exiting on unhandled exception: std::bad_alloc".

Reply:
@tgrabiec  So I suppose the max Tablets per table will be 8192, since otherwise our own tests can't be run. In that case the memory allocated by the `_tablet_to_dc_replication_factor_map` vector is 8192*8=65536 bytes (counting there are 4 data centers). Do you still think it makes sense to switch to a chunked vector (which allocates by default by 128k).

Reply:
>  So I suppose the max Tablets per table will be 8192

I think it's an order of magnitude or two below what we're aiming for.

Compare to vnodes -- the number of vnodes in 100 node cluster is by default 25600 (256 vnodes per node), then multiply it by number of shards per node, e.g. 64 -- becomes ~1.6*10^6  and I imagine we can reach similar number of tablets. More if cluster becomes larger.

Reply:
> In the benchmark I see that the number of tablets per table is 2048. Is that a typical number for production or it's just a testing number? Also when I set it to 65k I get an error "Exiting on unhandled exception: std::bad_alloc".

Can you share the command line which fails?

We aim at 100k tablets per cluster.

Some tests may not be able to handle that in a single shard due to overhead of tablet replica. In a real cluster, that overhead would be spread among many shards.

Reply:
@tgrabiec I run the following cmd 
`./build/dev/scylla perf-tablets  --tablets-per-table 65536`  
It fails with `std::bad_alloc`. In fact everything higher than 8192 is failing.

Reply:
> @tgrabiec I run the following cmd `./build/dev/scylla perf-tablets --tablets-per-table 65536` It fails with `std::bad_alloc`. In fact everything higher than 8192 is failing.

The default number of tables is 100, so with --tablets-per-table 65536 you have 6.5 M tablets. That's beyond our target of 100k.

tablet_map takes 600 MB. It probably bad_allocs due to mutation object size when persisting. With -m8G it proceeds, but then it takes almost a minute to save:

```
INFO  2024-11-06 14:06:48,099 [shard 0:stat] testlog - Saved in 41656.312500 [ms]
INFO  2024-11-06 14:07:05,748 [shard 0:stat] testlog - Read in 17534.886719 [ms]
```

And we fail in read_tablet_mutations() due to:

```
ERROR 2024-11-06 14:07:08,778 [shard 0:main] seastar - Exiting on unhandled exception: std::runtime_error (Maximum amount of memory for building query results is exhausted, unpaged query cannot be finished)
```

That one is used in migration_manager.cc, in schema pull handler:
```
            if (options->group0_snapshot_transfer) {
                cm.emplace_back(co_await db::system_keyspace::get_group0_history(db));
                if (proxy.local().local_db().get_config().enable_tablets()) {
                    for (auto&& m: co_await replica::read_tablet_mutations(db)) {
```

We shouldn't use that path anymore for snapshot transfer, right @kbr-scylla ?

Reply:
I opened https://github.com/scylladb/scylladb/issues/21456